rust/src/storage/file/mod.rs

```diff
@@ -58,25 +65,29 @@ impl StorageBackend for FileStorageBackend {
     }
 
     async fn put_obj(&self, path: &str, obj_bytes: &[u8]) -> Result<(), StorageError> {
-        let dir = std::path::Path::new(path);
+        let tmp_file_name = format!("{}.temporary", Uuid::new_v4().to_string());
+        let tmp_path = self.join_path(&self.root, &tmp_file_name);
 
-        if let Some(d) = dir.parent() {
-            fs::create_dir_all(d).await?;
+        // run this in loop in case tmp file with tmp_file_name already exists
+        loop {
+            let mut rf = fs::OpenOptions::new()
+                .create_new(true)
+                .write(true)
+                .open(&tmp_path)
+                .await;
+            match rf {
+                Ok(mut f) => {
+                    f.write(obj_bytes).await?;
+                    break;
+                },
+                Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {
+                    continue;
```

Comment:
we should also add a max retry, i don't think infinitely looping for uuid conflicts will happen in reality, but could happen due to bugs in our code.

Reply:
yeap, I'm just not sure how to actually test it since we need to mock UUID call somehow to reproduce an error

Reply:
i feel like a best effort on preventing infinite loop here is good enough, we don't have to write test for it. I also don't have a good idea on how to patch the return value of uuid here without changing the storage backend struct.